Cookbook/Proposals/Danger-Rule-Xcodeproj.md

```diff
@@ -0,0 +1,132 @@
+
+This can be done pretty easily with Ruby, conceptually this will look like the logic below:
+
+```ruby
+pbxprojs = git.modified_files.select { |f| f.end_with?('pbxproj') }
+pbxprojs.each do |filename|
+  File.foreach(filename).with_index do |line, line_num|
+    warn("null reference found in pbxproj", file: filename, line: line_num) if line.include?('/* (null) */') }
+    warn("recovered references found in pbxproj", file: filename, line: line_num) if line.include?('name = "Recovered References";') }
+  end
+end
+```
+
+### Invalid UUIDs
+
+This is about detecting orphan or invalid UUIDs in the projects, which are printed by to STDERR when parsing the project using `xcodeproj`:
+
+```
+<PBXGroup path=`Tests` UUID=`F26256AC222EC21500DD5E7B`>` attempted to initialize an object with an unknown UUID. `5FFC934E2296E0980088BFAA` for attribute: `children`. This can be the result of a merge and  the unknown UUID is being discarded.
+```
+
+We can intercept the warnings printed by `xcodeproj`, and transform them into warnings in the Danger comment, using this kind of snippet that we just have to declare before even using the `Xcodeproj::Project.open(...)` we use for all other rules:
+
+```ruby
+module Xcodeproj::UserInterface
+  def self.warn(message)
+    Danger.warn message
+  end  
+end
+```
+
+We could even go one step further and parse the warnings for those invalid UUIDs in order to then detect on which line that UUID is in the related `pbxproj` file and make the warning comment be inline.
+
+### Unexpected Resources in Target
+
+Using the `xcodeproj` gem (which is already part of our dependencies as it's used by `cocoapods`), we can also quickly analyze the structure of the project and detect resources that shouldn't have been added to build phases.
+
+The draft ruby code below would for example loop on all `pbxproj` files modified by the PR, and for each of those Xcode projects, loop on all their targets to check if the "Resources Build Phase" of each target contain any of the files we want to forbid there:
+
+```ruby
+pbxprojs.each do |project_file|
+  proj = Xcodeproj::Project.open(project_file)
+  proj.targets.each do |target|
+    rsrc_files = target.build_phases
+      .find { |p| p.is_a?(Xcodeproj::Project::Object::PBXResourcesBuildPhase) }
+      .files_references.map(&:path)
+    rsrc_files.each do |path|
+      fail("#{path} should be removed from #{project_file}:#{target.name}" if path.end_with? '.xcconfig'
+    end
+  end
+end
+```
+
+We can use this logic to warn about those type of files/folders added to targets' resource phases by mistake:
+ - `.xcconfig` files
+ - `Info.plist` 
+ - `__Snapshots__` directories
+
+### Detect tests not added to scheme
+
+We could once again use the `xcodeproj` gem to detect if any new tests file is properly added to a test target, and that the corresponding test is enabled in the test action in the scheme, to ensure that no tests are skipped.
```

Comment:
Team seem to agree we don't need this rule and it might trigger some false positives if we did (since things are not uniform rn) so don't implement this one.